---
title: "Chapter 8: Conditions"
subtitle: "Maya Gans"
author: "R4DS Reading Group"
output:
  xaringan::moon_reader:
    nature:
      highlightLines: true
    lib_dir: libs
    css: xaringan-themer.css
    includes:
      after_body: insert-logo.html
---


```{r xaringan-themer, include = FALSE}
library(xaringanthemer)
library(tidyverse)
library(rlang)
library(gt)
mono_light(
  base_color = "#3092FF",
  header_font_google = google_font("Josefin Slab"),
  text_font_google   = google_font("Raleway", "300", "300i")
)
```


# Overview

- Signaling conditions
  - Error
  - Warning
  - Message
- Ignoring conditions
 - try
 - suppress Warning/Message
- Handling conditions
 - tryCatch
 - withCallingHandlers

```{r, warning=FALSE, message=FALSE}
brewing_materials <- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-03-31/brewing_materials.csv')
beer_taxed <- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-03-31/beer_taxed.csv')
brewer_size <- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-03-31/brewer_size.csv')
beer_states <- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-03-31/beer_states.csv')
```

The condition system splits the responsibilities into three parts–signaling:
 1. a condition
 2. handling it
 3. restarting

---

class: inverse, hide-logo
# Signaling Conditions 

- Error
- Warning
- Message

---

# Errors

> "An error message should start with a general statement of the problem then give a concise description of what went wrong. Consistent use of punctuation and formatting makes errors easier to parse. 

> (This guide is currently almost entirely aspirational; most of the bad examples come from existing tidyverse code.)"

.footnote[
https://style.tidyverse.org/error-messages.html
]

---

## stop

```{r, eval=FALSE}
beer_mean_error <- function(x) {
  if (!is.numeric(x)) {
    stop("Need numeric column", call.=FALSE) #<<
    mean(which(!is.na(x)[x]))
  } else{
    mean(which(!is.na(x)[x]))
  }
}

beer_mean_error(beer_states$states)
```

```
Error: Need numeric column
   2. stop("Need numeric column", call. = FALSE)
   1. beer_mean(beer_states$state)
```

---

## `rlang::abort`

```{r, eval=FALSE}
beer_mean_abort <- function(x) {
  if (!is.numeric(x)) {
    abort( #<<
      message = "Need numeric column", #<<
      arg = x #<<
      ) #<<
    mean(which(!is.na(x)[x]))
  } else{
    mean(which(!is.na(x)[x]))
  }
}

beer_mean_abort(beer_states$state)
```

```
Error: Need numeric column
   4. stop(fallback)
   3. signal_abort(cnd)
   2. abort(message = "Need numeric column", arg = x)
   1. beer_mean(beer_states$state)
```

---

## abort + glue

```{r}
beer_mean_abort <- function(data, x) {
  
  column_name <- deparse(substitute(x))
  
  msg <- glue::glue("Can't calculate mean, {column_name} is not numeric")
  
  if (!is.numeric(data[[x]])) {
    abort(
      message = msg, #<<
      arg = column_name, #<<
      data = data #<<
    )
    mean(which(!is.na(data[[x]])[data[[x]]]))
  } else {
    mean(which(!is.na(data[[x]])[data[[x]]]))
  }
}
```

```{r, eval=FALSE}
beer_mean_abort(beer_states, "state")
```


```
Error: Can't calculate mean, "state" is not numeric
Run `rlang::last_error()` to see where the error occurred. 
   4. stop(fallback) 
   3. signal_abort(cnd) 
   2. abort(message = msg, arg = column_name, data = data) 
   1. beer_mean(beer_states, "state") 
```

---

# abort metadata

```{r, eval=FALSE}
str(catch_cnd(beer_mean_abort(beer_states, "state")))
```

```{r, eval=FALSE}
[1] "\"state\""
List of 5
 $ message: 'glue' chr "Can't calculate mean, \"state\" is not numeric" #<<
 $ trace  :List of 4
 ...
 $ parent : NULL
 $ arg    : chr "\"state\"" #<<
 $ data   : tibble [1,872 × 4] (S3: spec_tbl_df/tbl_df/tbl/data.frame) #<<
  ..$ state  : chr [1:1872] "AK" "AK" "AK" "AK" ...
  ..$ year   : num [1:1872] 2008 2009 2010 2011 2012 ...
  ..$ barrels: num [1:1872] 2068 2264 1929 2251 2312 ...
  ..$ type   : chr [1:1872] "On Premises" "On Premises" "On Premises" "On Premises" ...
  ..- attr(*, "spec")=
  .. .. cols(
  .. ..   state = col_character(),
  .. ..   year = col_double(),
  .. ..   barrels = col_double(),
  .. ..   type = col_character()
  .. .. )
 - attr(*, "class")= chr [1:3] "rlang_error" "error" "condition"
```

---

# Warnings

> Warnings occupy a somewhat challenging place between messages (“you should know about this”) and errors (“you must fix this!”)

```{r, echo=FALSE}
beer_states <- 
  beer_states %>% 
  mutate(year = lubridate::ymd(year, truncated = 2L))

is.Date <- function(x) {
  inherits(x, c("Date", "POSIXt"))
}
```

```{r, eval=FALSE}
beer_mean_warning <- function(data, x) {
  
  column_name <- deparse(substitute(x))
  
  if (is.Date(data[[x]])) {
    warning(glue::glue("Are you sure you wanna calculate the mean? {column_name} is of type date")) #<<
    mean(which(!is.na(data[[x]])[data[[x]]]))
  } else {
    mean(which(!is.na(data[[x]])[data[[x]]]))
  }
}

beer_mean(beer_states, "year")
```

```
[1] "2013-07-02"
Warning message:
In beer_mean(beer_states, "year") :
  Are you sure you wanna calculate the mean? "year" is of type date
```

.footnote[
_Not shown: (1) mutating the year column to a date type, (2) creating an `is.Date` function to check for date types. See .Rmd for code_
]

---

# Messages

>Good messages are a balancing act: you want to provide just enough information so the user knows what’s going on, but not so much that they’re overwhelmed.

```{r}
basic_summary_stats <- function(data, x, round_n = NULL, quiet = FALSE) {
  
  if (!is.numeric(data[[x]])) abort(glue::glue("Need numeric value to calculate stats and {x} is categorical"))
  if (is.null(round_n)) {
    
    if (isFALSE(quiet)) message("round_n argument null, rounding to 2 digits by default") #<<
    
    data %>% 
      summarise(
        missing = sum(is.na( data[[x]] )),
        mean = round(mean(which(!is.na(data[[x]])[data[[x]]])), 2)
       )
  } else {
    data %>% 
      summarise(
        missing = sum(is.na( data[[x]] )),
        mean = round(mean(which(!is.na(data[[x]])[data[[x]]])), round_n)
       )
  }
}

```

.pull-left[
```{r}
basic_summary_stats(beer_states, "barrels")
```
]

.pull-right[
```{r}
basic_summary_stats(beer_states, "barrels", quiet = TRUE)
```
]

---

class: inverse, hide-logo
# Ignoring Conditions

- try
- suppressMessage

---

# try

```{r, eval=FALSE}
beer_mean_try <- function(data, x) {
  try(data[[x]]) #<<
  NA
}

beer_mean_try(beer_states, "states")
```

```
[1] NA
```

---

# suppressMessages

```{r, eval=FALSE}
testthat::test_that("beer mean function works", {
  
  # calculate mean using base and round to two digits
  base_mean <- round(mean(which(!is.na(beer_states$barrels)[beer_states$barrels])), 2)
  
  # use our function and suppress warning since we're not supplying a rounding argument
  suppressMessages( #<<
    our_function <- basic_summary_stats(beer_states, "barrels") %>% pull(mean) #<<
  ) #<<
  
  # test that they are equal
  testthat::expect_equal(base_mean, our_function)
  # test that our function will produce a message
  testthat::expect_message(basic_summary_stats(beer_states, "barrels") %>% pull(mean))
})
```

.footnote[
_My thoughts behind this were that we don't want to clutter the output log when we run all our tests, but is this really best practice?_
]

---

class: inverse, hide-logo
# Handling Conditions 

- tryCatch
- withCallingHandlers

---

# tryCatch

```{r, eval=FALSE}
beer_mean_tryCatch <- function(expr) {
  tryCatch(
    error = function(cnd) NA, #<<
    {
      glue::glue(
        "Average Beer Barrels Produced: {round(expr,2)}" #<<
      )
    },
    finally = { print("Thank God for Beer!") } #<<
  )
}
```

.pull-left[
```{r, eval=FALSE}
beer_mean_tryCatch(beer_mean_abort(beer_states, "state"))
```

```
[1] "Thank God for Beer!"
[1] NA
```

]

.pull-right[
```{r, eval=FALSE}
beer_mean_tryCatch(beer_mean_abort(beer_states, "barrels"))
```

```
[1] "Thank God for Beer!"
Average Beer Barrels Produced: 806.46
```

]

.footnote[
_I find it interesting that the `finally` statement is printed **before** the code inside the tryCatch. Can anyone explain why?_
]

---

# tryCatch

_We can use tryCatch within the for loop to catch errors without breaking the loop_

```{r, eval=FALSE}
for (indx in 1:ncol(beer_states)) {
    tryCatch(
        expr = {
            basic_summary_stats(beer_states, names(beer_states[indx]))
            message("Iteration ", indx, " successful.")
        },
        error = function(e){
            message("* Caught an error on itertion ", indx)
            print(e)
        }
    )
}
```

```
* Caught an error on itertion 1
<error/rlang_error>
Need numeric value to calculate stats and state is categorical
Backtrace:
 1. base::tryCatch(...)
 5. global::basic_summary_stats(beer_states, names(beer_states[indx]))
* Caught an error on itertion 2
<error/rlang_error>
Need numeric value to calculate stats and year is categorical
Backtrace:
 1. base::tryCatch(...)
 5. global::basic_summary_stats(beer_states, names(beer_states[indx]))
round_n argument null, rounding to 2 digits by default
* Iteration 3 successful.
* Caught an error on itertion 4
<error/rlang_error>
Need numeric value to calculate stats and type is categorical
Backtrace:
 1. base::tryCatch(...)
 5. global::basic_summary_stats(beer_states, names(beer_states[indx]))
```

---

# withCallingHandlers

> R’s error handling system lets you separate the code that actually recovers from an error from the code that decides how to recover. Thus, you can put recovery code in low-level functions without committing to actually using any particular recovery strategy, leaving that decision to code in high-level functions.

---

```{r}
expensive_function <- function(x, 
                               # warning print the warning and send us to browser
                               warning = function(w) { print(paste('warning:', w ));  browser() },
                               # error print the error and send us to browser
                               error=function(e) { print(paste('e:',e )); browser()} ) {
  
  print(paste("big expensive step we don't want to repeat for x:",x))
  
  z <- x  # the "expensive operation"
  
  # second function on z that isn't expensive but could potentially error
  repeat
  # put code in here that actually handles the errors
  withRestarts(
    withRestarts(
      withCallingHandlers(
        {
          print(paste("attempt cheap operation for z:",z))
          return(log(z))
        },
        warning = warning,
        error = error
      ),
      # action restart will take
      # if the function fails this will take us to the browser
      # and findRestart will find this name
      force_positive = function() {z <<- -z} 
    ),
    # or we can use this is the browser
    set_to_one = function() {z <<- 1}
  )
}
```

.footnote[
http://www.win-vector.com/blog/2012/10/error-handling-in-r/?utm_source=rss&utm_medium=rss&utm_campaign=error-handling-in-r
]

---

# Using `browser` 

#### Success
```{r, eval=FALSE}
expensive_function(2)
```

```
[1] "big expensive step we don't want to repeat for x: 2"
[1] "attempt cheap operation for z: 2"
[1] 0.6931472
```

.pull-left[
#### Fail numeric

```{r, eval=FALSE}
expensive_function(-2)
```

```
[1] "big expensive step we don't want to repeat for x: -2"
[1] "attempt cheap operation for z: -2"
[1] "warning: simpleWarning in log(z): NaNs produced\n"
Called from: (function(w) { print(paste('warning:', w ));  
    browser() })(list(
    message = "NaNs produced", call = log(z)))
```

```{r, eval=FALSE}
Browse[1]> invokeRestart("force_positive")
```

```
[1] "attempt cheap operation for z: 2"
[1] 0.6931472
```
]

.pull-right[
#### Fail character

```{r, eval=FALSE}
expensive_function('a')
```

```
expensive_function('a')
[1] "big expensive step we don't want to repeat for x: a"
[1] "attempt cheap operation for z: a"
[1] "e: Error in log(z): non-numeric 
  argument to mathematical function\n"
  Called from: h(simpleError(msg, call))
```

```{r, eval=FALSE}
Browse[1]> invokeRestart("set_to_one")
```

```
[1] "attempt cheap operation for z: 1"
[1] 0
```
]

---

# Without `browser`

Get the handler functions to invoke the restart (rather than print error as in example above)

```{r}
force_positive <- function(w) {invokeRestart("force_positive")}
set_to_one <- function(e) {invokeRestart("set_to_one")}

auto_expensive_function = function(x) {
  expensive_function(x, warning=force_positive, error=set_to_one)
}
```

---


```{r}
auto_expensive_function(2)
```


```{r}
auto_expensive_function(-2)
```

```{r}
auto_expensive_function('a')
```

---

> We can use the condition system to allow a low-level function to detect a problem while parsing a log file and signal an error, to allow mid-level code to provide several possible ways of recovering from such an error, and to allow code at the highest level of the application to define a policy for choosing which recovery strategy to use.



.footnote[
http://adv-r.had.co.nz/beyond-exception-handling.html
]

<style>
pre {
  font-size: 11px;
}

.small-font {
  font-size: 13px;
}

.footnote {
  color: #800020;
  font-size: 9px;
}

</style>